using zip_iterators with host data, error with std::random_shuffle

I'm trying to write some test code using thrust::zip_iterator on code from the standard library:

``` c++


#include <thrust/iterator/zip_iterator.h>
#include <thrust/tuple.h>
#include <vector>
#include <thrust/sort.h>
#include <algorithm>
#include <iostream>

template<typename T, int sz> int size(T(&)[sz]) {
    return sz;
}

int main() {

    std::vector<int> x = {8,7,6,5};
    std::vector<int> y = {4,3,2,1};

    auto b = thrust::make_zip_iterator(thrust::make_tuple(x.begin(),y.begin()));
    auto e = thrust::make_zip_iterator(thrust::make_tuple(x.end(), y.end()));

    thrust::sort(b,e);

    //uncommenting either line causes an error
    //std::sort(b,e);
    //std::random_shuffle(b,e);

    for (unsigned int i = 0; i < x.size(); ++ i) {
        std::cerr << x[i] << "," << y[i] << "\n";
    }

    return 0;

}

```

I compiled with 

```
g++ -std=c++14 swap.cpp -o swap -Wall -I.  -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_CPP Wall -I.
```

If I uncomment either of the lines with std::sort or std::random_shuffle, I get an error. (I think both errors are similar...):

```
In file included from /usr/include/c++/5/bits/char_traits.h:39:0,
                 from /usr/include/c++/5/ios:40,
                 from /usr/include/c++/5/ostream:38,
                 from /usr/include/c++/5/iterator:64,
                 from ./thrust/iterator/iterator_traits.h:34,
                 from ./thrust/iterator/detail/zip_iterator_base.h:19,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
/usr/include/c++/5/bits/stl_algobase.h: In instantiation of ‘void std::iter_swap(_ForwardIterator1, _ForwardIterator2) [with _ForwardIterator1 = thrust::zip_iterator<thrust::tuple<__gnu_cxx::__normal_iterator<int*, std::vector<int> >, __gnu_cxx::__normal_iterator<int*, std::vector<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >; _ForwardIterator2 = thrust::zip_iterator<thrust::tuple<__gnu_cxx::__normal_iterator<int*, std::vector<int> >, __gnu_cxx::__normal_iterator<int*, std::vector<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >]’:
/usr/include/c++/5/bits/stl_algo.h:4450:22:   required from ‘void std::random_shuffle(_RAIter, _RAIter) [with _RAIter = thrust::zip_iterator<thrust::tuple<__gnu_cxx::__normal_iterator<int*, std::vector<int> >, __gnu_cxx::__normal_iterator<int*, std::vector<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >]’
swap.cpp:25:28:   required from here
/usr/include/c++/5/bits/stl_algobase.h:148:11: error: no matching function for call to ‘swap(thrust::iterator_facade<thrust::zip_iterator<thrust::tuple<__gnu_cxx::__normal_iterator<int*, std::vector<int> >, __gnu_cxx::__normal_iterator<int*, std::vector<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, thrust::system::cpp::detail::tag, thrust::random_access_traversal_tag, thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, long int>::reference, thrust::iterator_facade<thrust::zip_iterator<thrust::tuple<__gnu_cxx::__normal_iterator<int*, std::vector<int> >, __gnu_cxx::__normal_iterator<int*, std::vector<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, thrust::system::cpp::detail::tag, thrust::random_access_traversal_tag, thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, long int>::reference)’
       swap(*__a, *__b);
           ^
In file included from /usr/include/c++/5/bits/stl_iterator.h:65:0,
                 from /usr/include/c++/5/iterator:63,
                 from ./thrust/iterator/iterator_traits.h:34,
                 from ./thrust/iterator/detail/zip_iterator_base.h:19,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
/usr/include/c++/5/bits/move.h:176:5: note: candidate: void std::swap(_Tp&, _Tp&) [with _Tp = thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>] <near match>
     swap(_Tp& __a, _Tp& __b)
     ^
/usr/include/c++/5/bits/move.h:176:5: note:   conversion of argument 2 would be ill-formed:
In file included from /usr/include/c++/5/bits/char_traits.h:39:0,
                 from /usr/include/c++/5/ios:40,
                 from /usr/include/c++/5/ostream:38,
                 from /usr/include/c++/5/iterator:64,
                 from ./thrust/iterator/iterator_traits.h:34,
                 from ./thrust/iterator/detail/zip_iterator_base.h:19,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
/usr/include/c++/5/bits/stl_algobase.h:148:11: error: invalid initialization of non-const reference of type ‘thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>&’ from an rvalue of type ‘thrust::iterator_facade<thrust::zip_iterator<thrust::tuple<__gnu_cxx::__normal_iterator<int*, std::vector<int> >, __gnu_cxx::__normal_iterator<int*, std::vector<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, thrust::system::cpp::detail::tag, thrust::random_access_traversal_tag, thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, long int>::reference {aka thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>}’
       swap(*__a, *__b);
           ^
In file included from /usr/include/c++/5/bits/stl_iterator.h:65:0,
                 from /usr/include/c++/5/iterator:63,
                 from ./thrust/iterator/iterator_traits.h:34,
                 from ./thrust/iterator/detail/zip_iterator_base.h:19,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
/usr/include/c++/5/bits/move.h:195:5: note: candidate: template<class _Tp, long unsigned int _Nm> void std::swap(_Tp (&)[_Nm], _Tp (&)[_Nm])
     swap(_Tp (&__a)[_Nm], _Tp (&__b)[_Nm])
     ^
/usr/include/c++/5/bits/move.h:195:5: note:   template argument deduction/substitution failed:
In file included from /usr/include/c++/5/bits/char_traits.h:39:0,
                 from /usr/include/c++/5/ios:40,
                 from /usr/include/c++/5/ostream:38,
                 from /usr/include/c++/5/iterator:64,
                 from ./thrust/iterator/iterator_traits.h:34,
                 from ./thrust/iterator/detail/zip_iterator_base.h:19,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
/usr/include/c++/5/bits/stl_algobase.h:148:11: note:   mismatched types ‘_Tp [_Nm]’ and ‘thrust::iterator_facade<thrust::zip_iterator<thrust::tuple<__gnu_cxx::__normal_iterator<int*, std::vector<int> >, __gnu_cxx::__normal_iterator<int*, std::vector<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, thrust::system::cpp::detail::tag, thrust::random_access_traversal_tag, thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, long int>::reference {aka thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>}’
       swap(*__a, *__b);
           ^
In file included from /usr/include/c++/5/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/5/bits/char_traits.h:39,
                 from /usr/include/c++/5/ios:40,
                 from /usr/include/c++/5/ostream:38,
                 from /usr/include/c++/5/iterator:64,
                 from ./thrust/iterator/iterator_traits.h:34,
                 from ./thrust/iterator/detail/zip_iterator_base.h:19,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
/usr/include/c++/5/bits/stl_pair.h:254:5: note: candidate: template<class _T1, class _T2> void std::swap(std::pair<_T1, _T2>&, std::pair<_T1, _T2>&)
     swap(pair<_T1, _T2>& __x, pair<_T1, _T2>& __y)
     ^
/usr/include/c++/5/bits/stl_pair.h:254:5: note:   template argument deduction/substitution failed:
In file included from /usr/include/c++/5/bits/char_traits.h:39:0,
                 from /usr/include/c++/5/ios:40,
                 from /usr/include/c++/5/ostream:38,
                 from /usr/include/c++/5/iterator:64,
                 from ./thrust/iterator/iterator_traits.h:34,
                 from ./thrust/iterator/detail/zip_iterator_base.h:19,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
/usr/include/c++/5/bits/stl_algobase.h:148:11: note:   ‘thrust::iterator_facade<thrust::zip_iterator<thrust::tuple<__gnu_cxx::__normal_iterator<int*, std::vector<int> >, __gnu_cxx::__normal_iterator<int*, std::vector<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, thrust::system::cpp::detail::tag, thrust::random_access_traversal_tag, thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, long int>::reference {aka thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>}’ is not derived from ‘std::pair<_T1, _T2>’
       swap(*__a, *__b);
           ^
In file included from ./thrust/detail/contiguous_storage.h:155:0,
                 from ./thrust/detail/temporary_array.h:26,
                 from ./thrust/system/detail/generic/scan_by_key.inl:25,
                 from ./thrust/system/detail/generic/scan_by_key.h:143,
                 from ./thrust/detail/scan.inl:27,
                 from ./thrust/scan.h:1563,
                 from ./thrust/system/detail/generic/reduce_by_key.inl:35,
                 from ./thrust/system/detail/generic/reduce_by_key.h:88,
                 from ./thrust/detail/reduce.inl:26,
                 from ./thrust/reduce.h:784,
                 from ./thrust/system/detail/generic/find.inl:19,
                 from ./thrust/system/detail/generic/find.h:62,
                 from ./thrust/detail/find.inl:25,
                 from ./thrust/find.h:384,
                 from ./thrust/system/detail/generic/sort.inl:26,
                 from ./thrust/system/detail/generic/sort.h:153,
                 from ./thrust/detail/sort.inl:26,
                 from ./thrust/sort.h:1361,
                 from swap.cpp:5:
./thrust/detail/contiguous_storage.inl:263:8: note: candidate: template<class T, class Alloc> void thrust::swap(thrust::detail::contiguous_storage<T, Alloc>&, thrust::detail::contiguous_storage<T, Alloc>&)
   void swap(detail::contiguous_storage<T,Alloc> &lhs, detail::contiguous_storage<T,Alloc> &rhs)
        ^
./thrust/detail/contiguous_storage.inl:263:8: note:   template argument deduction/substitution failed:
In file included from /usr/include/c++/5/bits/char_traits.h:39:0,
                 from /usr/include/c++/5/ios:40,
                 from /usr/include/c++/5/ostream:38,
                 from /usr/include/c++/5/iterator:64,
                 from ./thrust/iterator/iterator_traits.h:34,
                 from ./thrust/iterator/detail/zip_iterator_base.h:19,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
/usr/include/c++/5/bits/stl_algobase.h:148:11: note:   ‘thrust::iterator_facade<thrust::zip_iterator<thrust::tuple<__gnu_cxx::__normal_iterator<int*, std::vector<int> >, __gnu_cxx::__normal_iterator<int*, std::vector<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, thrust::system::cpp::detail::tag, thrust::random_access_traversal_tag, thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, long int>::reference {aka thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>}’ is not derived from ‘thrust::detail::contiguous_storage<T, Alloc>’
       swap(*__a, *__b);
           ^
In file included from ./thrust/pair.h:282:0,
                 from ./thrust/tuple.h:35,
                 from ./thrust/iterator/detail/zip_iterator_base.h:24,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
./thrust/detail/pair.inl:127:10: note: candidate: template<class T1, class T2> void thrust::swap(thrust::pair<T1, T2>&, thrust::pair<T1, T2>&)
     void swap(pair<T1,T2> &x, pair<T1,T2> &y)
          ^
./thrust/detail/pair.inl:127:10: note:   template argument deduction/substitution failed:
In file included from /usr/include/c++/5/bits/char_traits.h:39:0,
                 from /usr/include/c++/5/ios:40,
                 from /usr/include/c++/5/ostream:38,
                 from /usr/include/c++/5/iterator:64,
                 from ./thrust/iterator/iterator_traits.h:34,
                 from ./thrust/iterator/detail/zip_iterator_base.h:19,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
/usr/include/c++/5/bits/stl_algobase.h:148:11: note:   ‘thrust::iterator_facade<thrust::zip_iterator<thrust::tuple<__gnu_cxx::__normal_iterator<int*, std::vector<int> >, __gnu_cxx::__normal_iterator<int*, std::vector<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, thrust::system::cpp::detail::tag, thrust::random_access_traversal_tag, thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, long int>::reference {aka thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>}’ is not derived from ‘thrust::pair<T1, T2>’
       swap(*__a, *__b);
           ^
In file included from ./thrust/tuple.h:34:0,
                 from ./thrust/iterator/detail/zip_iterator_base.h:24,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
./thrust/detail/tuple.inl:795:6: note: candidate: void thrust::swap(thrust::tuple<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9>&, thrust::tuple<U0, U1, U2, U3, U4, U5, U6, U7, U8, U9>&) [with T0 = int&; T1 = int&; T2 = thrust::null_type; T3 = thrust::null_type; T4 = thrust::null_type; T5 = thrust::null_type; T6 = thrust::null_type; T7 = thrust::null_type; T8 = thrust::null_type; T9 = thrust::null_type; U0 = int&; U1 = int&; U2 = thrust::null_type; U3 = thrust::null_type; U4 = thrust::null_type; U5 = thrust::null_type; U6 = thrust::null_type; U7 = thrust::null_type; U8 = thrust::null_type; U9 = thrust::null_type] <near match>
 void swap(thrust::tuple<T0,T1,T2,T3,T4,T5,T6,T7,T8,T9> &x,
      ^
./thrust/detail/tuple.inl:795:6: note:   conversion of argument 2 would be ill-formed:
In file included from /usr/include/c++/5/bits/char_traits.h:39:0,
                 from /usr/include/c++/5/ios:40,
                 from /usr/include/c++/5/ostream:38,
                 from /usr/include/c++/5/iterator:64,
                 from ./thrust/iterator/iterator_traits.h:34,
                 from ./thrust/iterator/detail/zip_iterator_base.h:19,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
/usr/include/c++/5/bits/stl_algobase.h:148:11: error: invalid initialization of non-const reference of type ‘thrust::tuple<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>&’ from an rvalue of type ‘thrust::tuple<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>’
       swap(*__a, *__b);
           ^
In file included from ./thrust/detail/tuple.inl:18:0,
                 from ./thrust/tuple.h:34,
                 from ./thrust/iterator/detail/zip_iterator_base.h:24,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
./thrust/detail/swap.h:28:13: note: candidate: void thrust::swap(Assignable1&, Assignable2&) [with Assignable1 = thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>; Assignable2 = thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>] <near match>
 inline void swap(Assignable1 &a, Assignable2 &b)
             ^
./thrust/detail/swap.h:28:13: note:   conversion of argument 2 would be ill-formed:
In file included from /usr/include/c++/5/bits/char_traits.h:39:0,
                 from /usr/include/c++/5/ios:40,
                 from /usr/include/c++/5/ostream:38,
                 from /usr/include/c++/5/iterator:64,
                 from ./thrust/iterator/iterator_traits.h:34,
                 from ./thrust/iterator/detail/zip_iterator_base.h:19,
                 from ./thrust/iterator/zip_iterator.h:35,
                 from swap.cpp:2:
/usr/include/c++/5/bits/stl_algobase.h:148:11: error: invalid initialization of non-const reference of type ‘thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>&’ from an rvalue of type ‘thrust::iterator_facade<thrust::zip_iterator<thrust::tuple<__gnu_cxx::__normal_iterator<int*, std::vector<int> >, __gnu_cxx::__normal_iterator<int*, std::vector<int> >, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type> >, thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, thrust::system::cpp::detail::tag, thrust::random_access_traversal_tag, thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, long int>::reference {aka thrust::detail::tuple_of_iterator_references<int&, int&, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>}’
       swap(*__a, *__b);
           ^

```

I can't find a replacement zip_iterator for c++ (e.g. the one in boost is hobbled compared to thrust), so I am trying to use the ones in thrust. I can work around the random_shuffle pretty easily, but is there a simple way to get thrust zip_iterators working with std functions like this in general?

--- a/thrust/detail/device_reference.inl
+++ b/thrust/detail/device_reference.inl
@@ -46,7 +46,7 @@ template<typename T>
 
 template<typename T>
 __host__ __device__
-void swap(device_reference<T> &a, device_reference<T> &b)
+void swap(device_reference<T> a, device_reference<T> b)
 {
   a.swap(b);
 } // end swap()

--- a/thrust/device_reference.h
+++ b/thrust/device_reference.h
@@ -958,7 +958,7 @@ template<typename T>
  */
 template<typename T>
 __host__ __device__
-void swap(device_reference<T> &x, device_reference<T> &y);
+void swap(device_reference<T> x, device_reference<T> y);
 
 // declare these methods for the purpose of Doxygenating them
 // they actually are defined for a derived-from class

--- a/thrust/iterator/detail/tuple_of_iterator_references.h
+++ b/thrust/iterator/detail/tuple_of_iterator_references.h
@@ -241,6 +241,20 @@ template<
 };
 
 
+// this overload of swap() permits swapping tuple_of_iterator_references returned as temporaries from
+// iterator dereferences
+template<
+  typename T0, typename T1, typename T2, typename T3, typename T4, typename T5, typename T6, typename T7, typename T8, typename T9,
+  typename U0, typename U1, typename U2, typename U3, typename U4, typename U5, typename U6, typename U7, typename U8, typename U9
+>
+inline __host__ __device__
+void swap(tuple_of_iterator_references<T0,T1,T2,T3,T4,T5,T6,T7,T8,T9> x,
+          tuple_of_iterator_references<U0,U1,U2,U3,U4,U5,U6,U7,U8,U9> y)
+{
+  x.swap(y);
+}
+
+
 } // end detail
 } // end thrust
 

